### PR TITLE
feat(health-check): health checks added for each container

### LINKED ISSUE #57: Extend Helm chart with default liveness & readiness probes
Some basic default readiness/lieness Probes would be may a good idea too (e.g. tcp socket check), better then nothing?

_Originally posted by @spielkind in https://github.com/mittwald/kube-httpcache/pull/51#discussion_r582214484_

### PR DESCRIPTION
**Pull request by bilalcaliskan**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it:**
This PR adds liveness/readiness probes for each possible container in a pod. It will make us ensure that our deployments are healthy all the time.

**Which issue(s) this PR fixes:**
Fixes #57 

**Special notes for your reviewer:**
I had to add **tcpSocket** health probes for signaller port(8090) because application normally opens a single port, without varnish-exporter:
```shell
root@kube-httpcache-0:/# netstat -lntpu
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp6       0      0 :::8090                 :::*                    LISTEN      1/kube-httpcache
```

Also i have no idea how to publish new Helm chart https://helm.mittwald.de Helm repository. I could not examine deeply but i hope current Github actions handle that already.

### CODE REVIEW

**Comment by bilalcaliskan** [comment]
@martin-helmich i do not know how to interact with bot assign someone for review(if any), so could you assist me for the review process?

Regards.

**Comment by Shanuson** [comment]
We are currently implementing an custom version of this in our environment.
Using a readiness probe resulted in a problem when we only have one pod running, or installing the initial setup.
Initially the readiness probe will prevent the varnish pod from being listed by the svc as an endpoint, the go-lang script will therefore find no endpoints, so the varnish does not start and the readiness probe never becomes green. 

If you already got an cluster running, doing a rolling update works because other pods are listed in the endpoint list.

To make this work, the golang script needs to be updated too so it includes always itself (own pod ip) as an endpoint. 
So there will always be an endpoint in the list and varnish will start and the readiness probe will be fine.

File: chart/values.yaml [hunk 1]
```diff
@@ -152,12 +152,20 @@ exporter:
...
-  #    port: 6083
-  readinessProbe: {}
+  livenessProbe:
+    httpGet:
+      path: /
+      port: 9131
+    initialDelaySeconds: 20
+    periodSeconds: 10
+    timeoutSeconds: 3
+  readinessProbe:
```
Comment (Shanuson) [inline chart/values.yaml:162]: see my comment about the effect of a readiness probe,
but I guess these here should work since they are not using the varnish itself

**Comment by bilalcaliskan** [comment]
@Shanuson let me clear about one thing. Readiness or liveness probe is not related with any of the objects like service or endpoint. Kubelet does the readiness or liveness probe with the ip of the pod itself. Am i wrong?

**Comment by bilalcaliskan** [comment]
@Shanuson as a counter argument, i did not have any problem with that readiness/liveness probes.

**Comment by martin-helmich** [comment]
Sorry for the late reaction.

Regarding the "readiness vs. endpoint list" discussion: We've had that before, most recently here https://github.com/mittwald/kube-httpcache/issues/49#issuecomment-737082131.

TL;DR: If you're using clustering (multiple Varnish servers, using the signaller component to dispatch merge requests), a readiness probe might break service discovery (which relies on a service selecting all kube-httpcache pods) and might break startup entirely, because the readiness of kube-httpcache depends on selecting at least one endpoint from its own cluster (meaning that the controller won't become "ready" until at least one controller is ready -- with the obvious conclusion that it'll wait infinitely for itself).

To mitigate, one might attempt using a Service resource with the [`.spec.publishNotReadyAddresses` property](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#servicespec-v1-core).

**Comment by bilalcaliskan** [comment]
hello @martin-helmich , i could not test the PR with multi instance so i could not see any problem. I got the real problem right now, so thank you for clear description. I will continue to work on the PR for a while.

Cheers!
